src/encrypt.rs

```diff
@@ -0,0 +1,221 @@
+                },
+                "aescbc" => Provider {
+                    aesgcm: None,
+                    aescbc: Some(AesCbc {
+                        keys: vec![Key {
+                            name: "1".to_string(),
+                            secret: key.clone(),
+                        }],
+                    }),
+                    identity: None,
+                },
+                "identity" => Provider {
+                    aesgcm: None,
+                    aescbc: None,
+                    identity: Some(Identity {}),
+                },
+                _ => {
+                    bail!("unsupported encryption type");
+                }
+            };
+            resource.providers.push(provider);
+            // Always put identity in the providers
+            resource.providers.push(Provider {
+                aesgcm: None,
+                aescbc: None,
+                identity: Some(Identity {}),
+            });
+            resources.push(resource);
+        }
+
+        Ok(Self {
+            apiVersion: API_VERSION.to_string(),
+            kind: KIND.to_string(),
+            resources,
+        })
+    }
+
+    pub(crate) async fn new_kube_apiserver_config(encryption_type: String) -> Result<Self> {
+        Self::new(vec!["configmaps".to_string(), "secrets".to_string()], encryption_type).await
+    }
+
+    pub(crate) async fn new_openshift_apiserver_config(encryption_type: String) -> Result<Self> {
+        Self::new(vec!["routes.route.openshift.io".to_string()], encryption_type).await
+    }
+
+    pub(crate) async fn new_oauth_apiserver_config(encryption_type: String) -> Result<Self> {
+        Self::new(
+            vec![
+                "oauthaccesstokens.oauth.openshift.io".to_string(),
+                "oauthauthorizetokens.oauth.openshift.io".to_string(),
+            ],
+            encryption_type,
+        )
+        .await
+    }
+
+    pub(crate) fn parse_from_file(config_bytes: Vec<u8>) -> Result<Self> {
+        let config: EncryptionConfiguration = serde_json::from_slice(&config_bytes)?;
+
+        ensure!(config.kind == KIND, format!("kind should equal {}, found {}", KIND, config.kind));
+        ensure!(
+            config.apiVersion == API_VERSION,
+            format!("apiVersion should equal {}, found {}", API_VERSION, config.apiVersion)
+        );
+
+        Ok(config)
+    }
+
+    pub(crate) fn remove_redundant_providers(&mut self) {
+        for res in &mut self.resources {
+            if res.providers.len() > 1 {
+                res.providers = vec![res.providers[0].clone()]
+            }
+        }
+    }
+}
+
+async fn generate_key() -> Result<Vec<u8>> {
+    let output = Command::new("openssl")
+        .arg("rand")
+        .arg("32")
+        .output()
+        .await
+        .context("failed to run openssl command")?;
+
+    ensure!(
+        output.status.success(),
+        format!("openssl command failed with status: {}", output.status)
+    );
+
```

Comment:
We should probably `ensure!` we're reading exactly 32 bytes and not some random CLI junk

Also move `32` (also used above) into a constant (const AES_KEY_SIZE 256, AES_KEY_SIZE_BYTES = AES_KEY_SIZE / 8) and use it in the ensure

Reply:
Once more, this makes total sense. Is this what you had in mind?